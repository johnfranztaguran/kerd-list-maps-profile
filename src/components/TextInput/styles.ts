import { StyleSheet, TextStyle, ViewStyle } from 'react-native';
import { colors } from 'constant';
import textStyle from '../Text/styles';

interface Styles {
  textInput: ViewStyle;
  infoContainer: TextStyle;
  rightContainer: ViewStyle;
  highlighted: ViewStyle;
  view: ViewStyle;
  info: ViewStyle;
  text: ViewStyle;
}

const styles: Styles = StyleSheet.create<Styles>({
  textInput: {
    height: 40,
  },
  infoContainer: {
    marginLeft: 16,
    marginRight: 16,
  },
  rightContainer: {
    position: 'absolute',
    top: 40,
    right: 10,
    zIndex: 1,
    marginRight: 16,
  },
  highlighted: {
    color: colors.java,
    textDecorationLine: 'underline',
  },
  view: {
    marginTop: -8,
    marginBottom: -10,
  },
  info: {
    ...textStyle.caption,
    marginTop: 10,
  },
  text: {
    ...textStyle.caption,
    color: colors.cinnabar,
  },
});

export default styles;
