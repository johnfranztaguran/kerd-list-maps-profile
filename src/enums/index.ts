export enum Screens {
  Main = 'Main',
  Login = 'Login',
  Register = 'Register',
  CompleteRegister = 'CompleteRegister',
  Welcome = 'Welcome',
  ListForm = 'ListForm',
  ProfileSettings = 'ProfileSettings',
  GoogleMaps = 'GoogleMaps',
  ListCrud = 'ListCrud',
  ProfileCrud = 'ProfileCrud'
}

export enum TextInputType {
  Number = 'number',
  Decimal = 'decimal',
  Text = 'text',
  Email = 'email',
  Password = 'password'
}

export enum MessageType {
  Error = 'error',
  Info = 'info'
}

export enum ErrorType {
  Empty = 'empty',
  Invalid = 'invalid',
  Existed = 'existed',
  Info = 'info',
  InvalidCombination = 'invalidCombination'
}

export enum HelperMessageType {
  Info = 'info',
  Error = 'error'
}

export enum ButtonMode {
  Outlined = 'outlined',
  Contained = 'contained',
  Text = 'text'
}

export enum ButtonType {
  PlainButton = 'button',
  ButtonIcon = 'button-icon'
}
