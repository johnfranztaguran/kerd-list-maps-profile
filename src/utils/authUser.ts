import React, { useEffect } from "react";
import { getAuth, onAuthStateChanged, User } from "firebase/auth";
import { useDispatch, useSelector } from 'react-redux';
import { setUserAuthV2 } from '@redux/actions/loginActions'

const auth = getAuth();

export function useAuth() {
  const dispatch: any = useDispatch();
  const [user, setUser] = React.useState<User>();

  useEffect(() => {
    const unsubscribeFromAuthStateChanged = onAuthStateChanged(auth, (user) => {
      if (user) {
        setUser(user);
        dispatch(setUserAuthV2(user))
      } else {
        setUser(undefined);
        dispatch(setUserAuthV2(undefined))
      }
    });
    return unsubscribeFromAuthStateChanged;
  }, []);

  return {
    user,
  };
}
