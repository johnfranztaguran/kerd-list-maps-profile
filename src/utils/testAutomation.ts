// testId - iOS , accessibilityLabel - Android
export const testIdProps = (id: string) => {
  const appiumTestId = id.toLowerCase().replace(/\s+/g, '-')
  return {
    accessible: true,
    testID: appiumTestId,
    accessibilityLabel: appiumTestId
  }
}
