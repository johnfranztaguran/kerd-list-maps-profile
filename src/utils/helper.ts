import { lensPath, set, anyPass, isEmpty, isNil } from 'ramda'
import numeral from 'numeral'
import { formats, regex, messages } from 'constant'
import { ErrorType } from 'enums'

type NumberOrString = number | string
export const createArrLensPath = (stringPath: string): NumberOrString[] => {
  const pathsArray: string[] = stringPath.split('.')
  const finalPathsArray: NumberOrString[] = pathsArray.reduce(
    (partial: NumberOrString[], next) => {
      if (next.includes('[') && next.includes(']')) {
        const field = next.split('[')[0] || ''
        const indexOfOpenBracket = next.indexOf('[') + 1
        const indexOfCloseBracket = next.indexOf(']')
        const idx = next.substring(indexOfOpenBracket, indexOfCloseBracket)
        return [...partial, field, parseInt(idx, 0)]
      }
      return [...partial, next]
    },
    []
  )

  return finalPathsArray
}

export const updateNode = <M>(path: string, value: any, model: M): M => {
  const pathArray = createArrLensPath(path)
  const lensObjPath = lensPath(pathArray)

  return set(lensObjPath, value, model)
}

export const formatNumber = (value: number, decimal?: number) => {
  return value.toFixed(decimal || 0).replace(regex.currency, '$1,')
}

export const formatNumberToFullMoneyString = (value: number) => {
  return formatNumber(value, 2)
}

export const formatNumberAsCompactString = (value: number) => {
  return numeral(value).format(formats.numeralsFormatOption.options.compact)
}

export const isNilOrEmpty = anyPass([isNil, isEmpty])

export const isNumeric = (value: string) => {
  return regex.numberString.test(value)
}

export const getErrorMessage = (field: string, errorType?: ErrorType) => {
  return errorType ? (messages[field][errorType] as string) : ''
}
