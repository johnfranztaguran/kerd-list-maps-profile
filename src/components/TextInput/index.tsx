/* eslint-disable @typescript-eslint/indent */
import React from 'react';
import {
  View,
  TouchableOpacity,
  TextInput as RNTextInput,
  TextInputProps,
} from 'react-native';
import { TextInput as RNPTextInput } from 'react-native-paper';
import Markdown from 'react-native-markdown-package';
import { testIdProps } from 'utils/testAutomation';
import { isNilOrEmpty } from 'utils/helper';
import { TextInputType, MessageType } from 'enums';
import { colors, strings } from 'constant';
import styles from './styles';

type Props = {
  id: string;
  inputType?:
    | TextInputType.Text
    | TextInputType.Number
    | TextInputType.Decimal
    | TextInputType.Email
    | TextInputType.Password;
  message?: string;
  messageType?: MessageType;
  withDecimal?: boolean;
  label?: string;
  placeholder?: string;
  value?: string;
  maxValue?: number;
  style?: any;
  autoFocus?: boolean;
  maxLength?: number;
  onValueChange: (id: string, value: any) => void;
  inputReference?: React.RefObject<RNTextInput>;
  secureTextEntry?: boolean;
  rightComponent?: JSX.Element;
  onPressInfo?: () => void;
  mode?: 'flat' | 'outlined';
  cursorColor?: string;
  theme?: object;
  infoContainerStyle?: object;
  onBlur?: (args: any) => void;
};

const TextInput = (props: Props): JSX.Element => {
  const {
    id,
    inputType = TextInputType.Text,
    messageType,
    message,
    label,
    style,
    placeholder,
    autoFocus,
    value,
    maxValue,
    maxLength,
    onValueChange,
    inputReference,
    secureTextEntry = false,
    rightComponent,
    onPressInfo,
    mode = 'outlined',
    cursorColor = colors.java,
    theme = {},
    infoContainerStyle = {},
    onBlur,
    ...otherProps
  } = props;

  const onTextChange = (text: string) => {
    onValueChange(id, text);
  };

  const onInputBlur = (args: any) => {
    if (isNilOrEmpty(value)) {
      onValueChange(id, '');
    }
    if (onBlur) {
      onBlur(args);
    }
  };

  const getKeyboardType = () => {
    switch (inputType) {
      case TextInputType.Number:
        return 'number-pad';
      case TextInputType.Decimal:
        return 'decimal-pad';
      default:
        return 'default';
    }
  };

  return (
    <View>
      <>
        <RNPTextInput
          mode={mode}
          autoFocus={autoFocus}
          label={label}
          error={messageType === MessageType.Error && !!message}
          placeholder={placeholder}
          value={value}
          onChangeText={onTextChange}
          keyboardType={getKeyboardType()}
          autoCapitalize="none"
          autoCorrect={false}
          onBlur={onInputBlur}
          ref={inputReference}
          style={[styles.textInput, style]}
          maxLength={maxLength}
          selectionColor={cursorColor}
          secureTextEntry={secureTextEntry}
          theme={theme}
          {...(otherProps as TextInputProps)}
          {...testIdProps(strings.testIdText.prefixInput + label)}
        />
        {rightComponent && (
          <View style={styles.rightContainer}>{rightComponent}</View>
        )}
      </>
      {messageType !== undefined && !!message && (
        <View style={[styles.infoContainer, infoContainerStyle]}>
          <TouchableOpacity
            {...testIdProps(`${strings.testIdText.prefixTxt} HelperText`)}
            activeOpacity={1.0}
            onPress={onPressInfo}>
            <Markdown
              styles={{
                // Markdown styles
                heading1: styles.info,
                view: styles.view,
                strong: styles.highlighted,
                text: styles.text,
              }}>
              {message}
            </Markdown>
          </TouchableOpacity>
        </View>
      )}
    </View>
  );
};

export default TextInput;
