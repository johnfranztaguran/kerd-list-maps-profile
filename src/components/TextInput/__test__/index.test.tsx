import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { act } from 'react-test-renderer';
import { MessageType, ErrorType } from 'enums';
import { getErrorMessage } from 'utils/helper';
import TextInput from '../index';

jest.mock('react-native-gesture-handler', () => {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const View = require('react-native/Libraries/Components/View/View');
  return {
    Swipeable: View,
    DrawerLayout: View,
    State: {},
    ScrollView: View,
    Slider: View,
    Switch: View,
    TextInput: View,
    ToolbarAndroid: View,
    ViewPagerAndroid: View,
    DrawerLayoutAndroid: View,
    WebView: View,
    NativeViewGestureHandler: View,
    TapGestureHandler: View,
    FlingGestureHandler: View,
    ForceTouchGestureHandler: View,
    LongPressGestureHandler: View,
    PanGestureHandler: View,
    PinchGestureHandler: View,
    RotationGestureHandler: View,
    /* Buttons */
    RawButton: View,
    BaseButton: View,
    RectButton: View,
    BorderlessButton: View,
    /* Other */
    FlatList: View,
    gestureHandlerRootHOC: jest.fn(),
    Direction: {},
  };
});

describe('Component: TextInput', () => {
  const onChange = jest.fn();
  const onFocus = jest.fn();
  const onBlur = jest.fn();
  it('Should render a TextInput', () => {
    const { baseElement, getByTestId } = render(
      <TextInput onValueChange={onChange} label="name" id="name" />,
    );

    const input = getByTestId('input-name');
    act(() => {
      fireEvent.changeText(input, 'test');
    });
    expect(input.props.value).toBe('test');

    act(() => {
      fireEvent.focus(input, onFocus());
    });
    expect(onFocus).toBeCalled();

    act(() => {
      fireEvent.focus(input, onBlur());
    });
    expect(onBlur).toBeCalled();

    expect(baseElement).toMatchSnapshot();
  });

  it('Should render a TextInput with error message', () => {
    const { baseElement } = render(
      <TextInput
        id="email"
        onValueChange={onChange}
        messageType={MessageType.Error}
        message={getErrorMessage('email', ErrorType.Empty)}
      />,
    );

    expect(baseElement).toMatchSnapshot();
  });
});
