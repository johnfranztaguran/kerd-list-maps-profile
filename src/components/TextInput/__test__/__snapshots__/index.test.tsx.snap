// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: TextInput Should render a TextInput 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          Array [
            undefined,
            Object {},
          ]
        }
      >
        <View>
          <View
            pointerEvents="none"
            style={
              Array [
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 6,
                },
                Object {
                  "backgroundColor": "#f6f6f6",
                  "borderColor": "#6200ee",
                  "borderRadius": 4,
                  "borderWidth": 2,
                },
              ]
            }
          />
          <View
            style={
              Object {
                "minHeight": 40,
                "paddingBottom": 0,
                "paddingTop": 8,
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                Object {
                  "bottom": 0,
                  "left": 0,
                  "opacity": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                  ],
                }
              }
            >
              <View
                pointerEvents="none"
                style={
                  Object {
                    "backgroundColor": "#f6f6f6",
                    "bottom": 0,
                    "height": 2,
                    "left": 10,
                    "opacity": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 6,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                    ],
                    "width": 8,
                  }
                }
              />
              <Text
                numberOfLines={1}
                style={
                  Object {
                    "backgroundColor": "#f6f6f6",
                    "color": "transparent",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "left": 18,
                    "opacity": 0,
                    "paddingHorizontal": 0,
                    "position": "absolute",
                    "textAlign": "left",
                    "top": 29,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateY": 0,
                      },
                      Object {
                        "scale": 1,
                      },
                      Object {
                        "scaleY": 1,
                      },
                    ],
                    "writingDirection": "ltr",
                  }
                }
              >
                name
              </Text>
              <Text
                numberOfLines={1}
                style={
                  Object {
                    "color": "#6200ee",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "left": 0,
                    "opacity": 0,
                    "paddingHorizontal": 14,
                    "position": "absolute",
                    "textAlign": "left",
                    "top": 28,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateY": 0,
                      },
                      Object {
                        "scale": 1,
                      },
                    ],
                    "writingDirection": "ltr",
                  }
                }
              >
                name
              </Text>
              <Text
                numberOfLines={1}
                style={
                  Object {
                    "color": "rgba(0, 0, 0, 0.54)",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "left": 0,
                    "opacity": 1,
                    "paddingHorizontal": 14,
                    "position": "absolute",
                    "textAlign": "left",
                    "top": 28,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateY": 0,
                      },
                      Object {
                        "scale": 1,
                      },
                    ],
                    "writingDirection": "ltr",
                  }
                }
              >
                name
              </Text>
            </View>
            <TextInput
              accessibilityLabel="input-name"
              accessible={true}
              allowFontScaling={true}
              autoCapitalize="none"
              autoCorrect={false}
              editable={true}
              keyboardType="default"
              multiline={false}
              placeholder=""
              placeholderTextColor="rgba(0, 0, 0, 0.54)"
              rejectResponderTermination={true}
              secureTextEntry={false}
              selectionColor="#17B2BF"
              style={
                Array [
                  Object {
                    "flexGrow": 1,
                    "margin": 0,
                    "paddingHorizontal": 14,
                    "textAlign": "left",
                    "zIndex": 1,
                  },
                  Object {
                    "height": 40,
                  },
                  Object {
                    "paddingBottom": 12,
                    "paddingTop": 12,
                  },
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "textAlignVertical": "center",
                  },
                ]
              }
              testID="input-name"
              underlineColorAndroid="transparent"
              value="test"
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Component: TextInput Should render a TextInput with error message 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          Array [
            undefined,
            Object {},
          ]
        }
      >
        <View>
          <View
            pointerEvents="none"
            style={
              Array [
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 6,
                },
                Object {
                  "backgroundColor": "#f6f6f6",
                  "borderColor": "#B00020",
                  "borderRadius": 4,
                  "borderWidth": 2,
                },
              ]
            }
          />
          <View
            style={
              Object {
                "minHeight": 40,
                "paddingBottom": 0,
                "paddingTop": 8,
              }
            }
          >
            <TextInput
              accessibilityLabel="input-undefined"
              accessible={true}
              allowFontScaling={true}
              autoCapitalize="none"
              autoCorrect={false}
              editable={true}
              keyboardType="default"
              multiline={false}
              placeholderTextColor="rgba(0, 0, 0, 0.54)"
              rejectResponderTermination={true}
              secureTextEntry={false}
              selectionColor="#17B2BF"
              style={
                Array [
                  Object {
                    "flexGrow": 1,
                    "margin": 0,
                    "paddingHorizontal": 14,
                    "textAlign": "left",
                    "zIndex": 1,
                  },
                  Object {
                    "height": 40,
                  },
                  Object {
                    "paddingBottom": 12,
                    "paddingTop": 12,
                  },
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "textAlignVertical": "center",
                  },
                ]
              }
              testID="input-undefined"
              underlineColorAndroid="transparent"
            />
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "marginLeft": 16,
              "marginRight": 16,
            },
            Object {},
          ]
        }
      >
        <TouchableOpacity
          accessibilityLabel="txt--helpertext"
          accessible={true}
          activeOpacity={1}
          testID="txt--helpertext"
        >
          <View
            style={
              Array [
                Object {
                  "alignSelf": "stretch",
                },
                Object {
                  "marginBottom": -10,
                  "marginTop": -8,
                },
              ]
            }
          >
            <Text
              style={
                Object {
                  "alignItems": "flex-start",
                  "flexDirection": "row",
                  "flexWrap": "wrap",
                  "justifyContent": "flex-start",
                  "marginBottom": 10,
                  "marginTop": 10,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#E74C3C",
                    "fontFamily": "OpenSans",
                    "fontSize": 10,
                    "lineHeight": 12,
                  }
                }
              >
                Please enter your email address.
              </Text>
            </Text>
          </View>
        </TouchableOpacity>
      </View>
    </View>
  </View>
</View>
`;
